File: packages/core/src/__tests__/propagation.test.ts
import { describe, it, expect, beforeEach } from 'vitest'
import { Engine } from '../engine.js'

describe('Propagation', () => {
  let engine: Engine

  beforeEach(() => {
    engine = new Engine({ maxPropagationRounds: 50 })
  })

  it('should propagate through a linear chain via nested emits', () => {
    const a = engine.event<number>('a')
    const b = engine.event<number>('b')
    const c = engine.event<number>('c')
    const results: number[] = []

    engine.on(a, (x) => engine.emit(b, x * 2))
    engine.on(b, (x) => engine.emit(c, x + 1))
    engine.on(c, (v) => results.push(v))

    engine.emit(a, 3)
    expect(results).toEqual([7]) // 3*2+1
  })

  it('should handle fan-out (one event, multiple consumers)', () => {
    const src = engine.event<number>('src')
    const out1 = engine.event<number>('out1')
    const out2 = engine.event<number>('out2')
    const r1: number[] = []
    const r2: number[] = []

    engine.on(src, (x) => engine.emit(out1, x + 10))
    engine.on(src, (x) => engine.emit(out2, x + 20))
    engine.on(out1, (v) => r1.push(v))
    engine.on(out2, (v) => r2.push(v))

    engine.emit(src, 1)
    expect(r1).toEqual([11])
    expect(r2).toEqual([21])
  })

  it('should handle diamond-shaped DAG with join', () => {
    const a = engine.event<number>('a')
    const left = engine.event<number>('left')
    const right = engine.event<number>('right')
    const results: number[] = []

    engine.on(a, (x) => engine.emit(left, x * 2))
    engine.on(a, (x) => engine.emit(right, x * 3))
    engine.on([left, right], (l: number, r: number) => {
      results.push(l + r)
    })

    engine.emit(a, 10)
    expect(results).toEqual([50]) // 10*2 + 10*3
  })

  it('should throw on excessive propagation rounds from nested emits', () => {
    const smallEngine = new Engine({ maxPropagationRounds: 5 })
    const a = smallEngine.event<number>('a')

    // Create a self-feeding loop via nested emit
    smallEngine.on(a, (val) => {
      smallEngine.emit(a, val + 1)
    })

    expect(() => smallEngine.emit(a, 0)).toThrow(/Propagation exceeded/)
  })

  it('should process events deposited during action execution', () => {
    const trigger = engine.event<number>('trigger')
    const sideEffect = engine.event<string>('sideEffect')
    const results: string[] = []

    engine.on(trigger, (val) => {
      engine.emit(sideEffect, `from-trigger-${val}`)
    })
    engine.on(sideEffect, (msg) => results.push(msg))

    engine.emit(trigger, 42)
    expect(results).toEqual(['from-trigger-42'])
  })

  it('should handle events with no consumers gracefully', () => {
    const orphan = engine.event<number>('orphan')
    expect(() => engine.emit(orphan, 99)).not.toThrow()
  })

  it('should handle multiple sequential emits independently', () => {
    const ev = engine.event<number>('ev')
    const results: number[] = []

    engine.on(ev, (v) => results.push(v))

    engine.emit(ev, 1)
    engine.emit(ev, 2)
    engine.emit(ev, 3)

    expect(results).toEqual([1, 2, 3])
  })

  it('should maintain synchronous propagation within a single emit call', () => {
    const a = engine.event<number>('a')
    const b = engine.event<number>('b')
    const c = engine.event<number>('c')
    const trace: string[] = []

    engine.on(a, (x) => {
      trace.push(`a:${x}`)
      engine.emit(b, x + 1)
    })
    engine.on(b, (x) => {
      trace.push(`b:${x}`)
      engine.emit(c, x + 1)
    })
    engine.on(c, (x) => {
      trace.push(`c:${x}`)
    })

    engine.emit(a, 1)
    // Everything should have been processed synchronously
    expect(trace).toEqual(['a:1', 'b:2', 'c:3'])
  })
})

File: packages/dart/test/propagation_test.dart
import 'package:test/test.dart';
import 'package:pulse/pulse.dart';

void main() {
  late Engine engine;

  setUp(() {
    engine = Engine(EngineOptions(maxPropagationRounds: 50));
  });

  test('should propagate through a linear chain', () {
    final a = engine.event<int>('a');
    final b = engine.event<int>('b');
    final c = engine.event<int>('c');
    final results = <int>[];

    engine.pipe<int, int>(a, b, (x) => x * 2);
    engine.pipe<int, int>(b, c, (x) => x + 1);
    engine.on<int>(c, (v) => results.add(v));

    engine.emit(a, 3);
    expect(results, equals([7])); // 3*2+1
  });

  test('should handle fan-out (one event, multiple consumers)', () {
    final src = engine.event<int>('src');
    final out1 = engine.event<int>('out1');
    final out2 = engine.event<int>('out2');
    final r1 = <int>[];
    final r2 = <int>[];

    engine.pipe<int, int>(src, out1, (x) => x + 10);
    engine.pipe<int, int>(src, out2, (x) => x + 20);
    engine.on<int>(out1, (v) => r1.add(v));
    engine.on<int>(out2, (v) => r2.add(v));

    engine.emit(src, 1);
    expect(r1, equals([11]));
    expect(r2, equals([21]));
  });

  test('should handle diamond-shaped DAG', () {
    final a = engine.event<int>('a');
    final b = engine.event<int>('b');
    final c = engine.event<int>('c');
    final d = engine.event<int>('d');
    final results = <int>[];

    engine.pipe<int, int>(a, b, (x) => x + 1);
    engine.pipe<int, int>(a, c, (x) => x + 2);
    engine.join<int>(
      inputs: [b, c],
      output: d,
      action: (payloads) => (payloads[0] as int) + (payloads[1] as int),
    );
    engine.on<int>(d, (v) => results.add(v));

    engine.emit(a, 10);
    expect(results, equals([23])); // (10+1)+(10+2)
  });

  test('should detect cycle in DAG when rules form a cycle', () {
    final a = engine.event<int>('a');
    final b = engine.event<int>('b');

    engine.pipe<int, int>(a, b, (x) => x + 1);
    expect(() {
      engine.pipe<int, int>(b, a, (x) => x + 1);
      engine.emit(a, 0);
    }, throwsA(predicate((e) =>
        e is StateError && e.message.contains('Cycle detected'))));
  });

  test('should throw on excessive propagation rounds from nested emits', () {
    final smallEngine = Engine(EngineOptions(maxPropagationRounds: 5));
    final a = smallEngine.event<int>('a');

    smallEngine.on<int>(a, (val) {
      smallEngine.emit(a, val + 1);
    });

    expect(
        () => smallEngine.emit(a, 0),
        throwsA(predicate(
            (e) => e is StateError && e.message.contains('Propagation exceeded'))));
  });

  test('should process events deposited during action execution', () {
    final trigger = engine.event<int>('trigger');
    final sideEffect = engine.event<String>('sideEffect');
    final results = <String>[];

    engine.on<int>(trigger, (val) {
      engine.emit(sideEffect, 'from-trigger-$val');
    });
    engine.on<String>(sideEffect, (msg) => results.add(msg));

    engine.emit(trigger, 42);
    expect(results, equals(['from-trigger-42']));
  });

  test('should handle events with no consumers gracefully', () {
    final orphan = engine.event<int>('orphan');
    expect(() => engine.emit(orphan, 99), returnsNormally);
  });

  test('should maintain topological order across fan-in', () {
    final a = engine.event<int>('a');
    final b = engine.event<int>('b');
    final c = engine.event<int>('c');
    final d = engine.event<int>('d');
    final trace = <String>[];

    engine.pipe<int, int>(a, b, (x) {
      trace.add('a->b');
      return x;
    });
    engine.pipe<int, int>(a, c, (x) {
      trace.add('a->c');
      return x;
    });
    engine.join<int>(
      inputs: [b, c],
      output: d,
      action: (payloads) {
        trace.add('join->d');
        return (payloads[0] as int) + (payloads[1] as int);
      },
    );

    engine.emit(a, 1);
    final joinIdx = trace.indexOf('join->d');
    final abIdx = trace.indexOf('a->b');
    final acIdx = trace.indexOf('a->c');
    expect(joinIdx, greaterThan(abIdx));
    expect(joinIdx, greaterThan(acIdx));
  });

  test('should handle multiple sequential emits independently', () {
    final ev = engine.event<int>('ev');
    final results = <int>[];

    engine.on<int>(ev, (v) => results.add(v));

    engine.emit(ev, 1);
    engine.emit(ev, 2);
    engine.emit(ev, 3);

    expect(results, equals([1, 2, 3]));
  });
}
